doc/configurator.md
# Configurator

Since the *Headless TX* does not have a user interface, we need an external device to configure it. We call this device or software *Configurator*.

When connected to a *Headless TX*, the *Configurator* can provide the following functions:

* Live configuration of inputs, mixers, sub-trim values, end points, etc
* Live configuration of the transmitter hardware (inputs, their names and types, trims, invert)
* Live mixer configuration (expo, dual-rate, 4-wheel steering, elevons ...)
* Live output configuration (reversing, endpoints, sub-trim ...)
* Calibration function for analog inputs (sticks, pots)
* Live display of stick, pot, switch, push-button, trims and output channel data
* Show the battery state

The *Configurator* can also provide other functionality:

* Maintain a database of models and transmitter configurations
* Synchronize the model and transmitter database with other *configurators* (not implemented yet)


The *Configurator* functionality can be implemented in different ways:

* Custom device

    Dedicated embedded device with display and buttons. Similar to a programming box for an ESC.

* Web app

    An application that runs in a Web browser

* Smartphone app


Using the a *Web app* as *Configurator*, which works in any modern browser, is the most attractive option as it can be used on a wide range of devices: PC, Smartphone, Tablet. Smartphones even allow web-apps to be installed as icon on the home screen and run full-screen, so we get a app-like user experience.



## Principle of operation

The *Headless TX* performs its operation based on a configuration that is stored in its persistent memory (Flash memory of the microcontroller). The configuration determines which inputs provides what functions, how the inputs are mixed to form output channels, etc. Please refere to the [architecture documentation](architecture.md) for details.

The firmware is designed such that the configuration can be changed at run-time; the *Headless TX* adapts to configurations dynamically.

The configuration comprises of settings that are *hardware specific*, and setting that are *model specific*. While the hardware and model specific settings are kept separate, they are stored in one area of consecutive memory.

The *Headless TX* provides an API that allows an external device -- the *Configurator* -- to change the configuration.

The API is very simple: Read parts of the configuration memory, write parts of the configuration memory, and copy from one memory area to another memory area within the configuration memory.

The advantage of this method is that the implementation in the *Headless TX* is very simple. On the down-side, the *Configurator* must know the precise memory layout of the all the elements within the configuration memory.



## Transports

The *Configurator* must be able to connect to a *Headless TX* in order to perform operations on the API provided by the *Headless TX*.

The *Headless TX* offers the following transports natively:

* nRF Enhanced Shockburst protocol

    The nRF protocol makes use of the nRF24L01 used to send data to the model. *Configurator* data is sent in pauses when no model transmission is taking place.

* UART protocol

    The UART (serial port) protocol is available by connecting to the UART provided by the microcontroller. The UART protocol is currently not implemented on the *Headless TX*, but used internally in the *nrF-to-Websocket bridge* described below.


The design goal is to keep the protocol running over the different transports the same, so that the transports appear transparent to the *Configurator*.


### Bridging

Unfortunately a *Web app* is not able to utilize any of the transports provided by the *Headless TX* directly.

To be able to make a *Configurator Web app*, we need a bridging mechanism that goes between the *Headless TX* and the *Configurator* running in the web browser.

On the web browser side, we have **HTTP** and **Websocket** protocols to our disposals.

Another potential transport would be **BLE**. The *Web Bluetooth* protocol allows web apps to connect to BLE devices. However, *Web Bluetooth* is relatively new and not mature yet.

From a physical implementation point-of-view, the bridge can be a separate box, or integrated into the *Headless TX*.
We are currently using a separate box as bridge, which can be used for multiple *Headless TX*. On the down-side, now we have to bring yet another (small) box along while playing with our RC toys.


### Bridge protocol considerations

Ideally the protocol that a *Headless TX bridge* provides to the *Configurator* is equivalent, or at least very similar, to the native transport protocols supported by the *Headless TX*.

### Websocket to nRF/UART bridge

The cheapest way to build a bridge with a HTTP/Websocket interface is by using **ESP8266** modules. The older ESP8266 modules have only a UART interface and not a lot of memory, but newer versions provide access to an SPI port that would allow direct interfacing with a nRF24L01 module.

In order to relief the ESP8266 firmware from the real-time requirements of dealing with the nRF24L01 module, the current bridge implementation uses a nRF51822 module with integrated ARM microcontroller to perform the real-time nRF protocol, sending and receiving information over UART to the ESP8266.

The current *Websocket to nRF bridge* is powered from a single, small Li-Ion cell.

In order to minimize interference between Wi-Fi and the nRF protocol, we run Wi-Fi on Channel 11. the nRF protocol uses mostly the lower channels of the 2.4 GHz ISM band.

The *Websocket to nRF bridge* can also easily be simulated using NodeJS. This allows development of the *Web app* without using actual hardware.



## Security considerations

While this device falls in the category of *Toy*, we still must consider potential security issues.

When the *Configurator* connects via the UART directly, the *Headless TX* can trust the *Configurator* -- after all, physical access to the *Headless TX* was obtained.

Things are different when the *Configurator* connects via one of the wireless methods. In case of an open system, anyone with a wireless *Configurator* could connect to our transmitter and ruin our day. This may not even be malicious: Other *Headless TX* users may accidentally connect to our transmitter while trying to configure their own unit.

A simple 4-digit pass-phrase can prevent accidental configuration from happening.

Since the pass-phrase is stored as part of the transmitter configuration, our own *Configurator* could automatically log into the transmitter as it can retrieve the pass-phrase from its local database. Others would not have the configuration of our transmitter in their database, so they would need to enter the pass-phrase.

Do note: a malicious user could sniff the pass-phrase while we connect to our transmitter, or obtain it from a sync'ed copy of our database. But then, a malicious user can also sniff our address and bind data and directly manipulate our model; or simply jam the 2.4 GHz band. We therefore consider the pass-phrase being sufficient.

Following tradition, the default pass-phrase of a virgin transmitter is the same as we use on our luggage: '1234'.



## nRF protocol details

The nRF protocol integrates with the RF protocol that is employed in the HobbyKing HK300, HK310 and 3XS transmitters. We refer to this protocol as *HK310 protocol* from now on.


### The *HK310 protocol*

The *HK310 protocol* has the following properties:

* The legacy nRF24 protocol (non Shockburst version) is used at 250 KBps
* The address is 5 bytes long and is unique for each model
* All packets sent by the transmitter have a payload length of 10 bytes
* One burst of transmission occurs every 5 ms
* A transmission burst comprises of:
    - Two identical stick (or failsafe) packets, with data for 3 channels
    - A bind packet, sent at low power with a fixed address on channel 81
* After each transmission burst, the the frequency is changed to the next in a list of 20 hop channels

One transmission burst takes about 1.2 ms.


    __MMMM_MMMM_BBBB________________________________MMMM_MMMM_BBBB_________

      <------------------- 5 ms ------------------->

    MMMM: Stick or Failsafe data, sent on the hop channel
    BBBB: Bind packet, sent on channel 81


One challenge in the *Headless TX* was that, due the use of inexpensive nRF24+PA (power amplifier) module, we were unable to output at a really low RF power. Even at the lowest power setting the bind packets can be received from quite far away. In order to avoid interference with other RC transmitters, the *Headless TX* therefore only outputs bind packets for the first 10 seconds after power on.


### Integrating with the *HK310 protocol*

We extend the transmission burst of the *HK310 protocol* with one more *configurator packet*. This implies that one *configurator packet* is sent every 5 ms.


    __MMMM_MMMM_BBBB_CCC____________________________MMMM_MMMM_BBBB_CCC_____

      <------------------- 5 ms ------------------->

    MMMM: Stick or Failsafe data, sent on the hop channel
    BBBB: Bind packet, sent on channel 81
    CCC: Configurator packet at 2 Mbps


The *configurator packet* use a different radio setup than the *HK310 protocol*:

* Enhanced Shockburst with auto-acknowledge (AA) enabled, with optional ACK payload
* 2 MBps data rate
* Variable payload length up to the maximum of 32 bytes
* 2-Byte CRC
* Lowest possible transmit power (at least on the *Headless TX* side, which utilizes a RF amplifier)
* ARD is set to 500 us

The *Headless TX* acts as the PTX (Primary role Transmitter, as defined in the nRF24 specification), the *Configurator* acts as PRX (Primary role Receiver).

The *Configurator* is able to send information to the Headless TX as part of the Acknowledge procedure when receiving a packet from the *Headless TX*.

From a protocol point-of-view, one could say that the *Headless TX* is the master, and the *Configurator* is the slave, with the master polling the slave in regular intervals.


### Overcoming reliability issues

Wireless protocols are inherent to interference and poor reception, resulting in lost packets.

To make it easier for users of the nRF protocol, the low-level implementation shall implement a two-byte packet buffer and inspect packet contents.

The nRF protocol is designed in such a way that every request by the *configurator* is answered by the *Headless TX* in the next packet. For example, a `CFG_READ` packet is answered with `TX_REQUESTED_DATA`; `CFG_WRITE` is answered with `TX_WRITE_SUCCESSFUL`.

In case a transmission error occurs, the next successful packet received from the *Headless TX* will not match the request sent, so the low-level implementation knows that something went wrong and can resend the packet.

A two packet buffer is required to achieve maximum throughput, as the nRF Shockburst protocol requires us to queue data before we can receive the answer to the previous request.

The above algorithm implies that packets may be executed out of order.


### nRF protocol states

The nRF Shockburst protocol is designed for one pair of PTX and PRX. Since there is a possibility that more than one *Configurator* is in use at the same time, we must ensure that integrity is preserved. To achieve this, we only allow one *Configurator* to connect to a *Headless TX* at a given moment in time.

From a protocol view, we achieve this by differentiating between *NOT CONNECTED* and *CONNECTED* state.


### nRF protocol *NOT CONNECTED* state

This is the default state for both the *Headless TX* and the *Configurator*.

All communication in this state happens on the fixed nRF channel **79** with address **4C:42:72:63:78** (`LBrcx` in hex).

The *Headless TX* sends a `TX_FREE_TO_CONNECT` packet whenever the transmission burst is transmitting on the first model hop channel. This means that `TX_FREE_TO_CONNECT` packets are sent every 100 ms (20 hop channels times 5 ms each).

Since there may be multiple *Headless TX* on the air, the `TX_FREE_TO_CONNECT` packet contains the 8-byte UUID of the transmitter as well as a human-readable transmitter name so that we can differentiate between transmitters.

The *Configurator* listens for `TX_FREE_TO_CONNECT` and shows the user a list of all transmitters in the neighborhood. When the user chooses a transmitter, the *Configurator* replies with `CFG_REQUEST_TO_CONNECT` and changes its state to `CONNECTED`.
The `CFG_REQUEST_TO_CONNECT` packet contains the UUID of the transmitter to connect to, the pass-phrase, a unique 5-byte address and hop channel information.

When the *Headless TX* receives a `CFG_REQUEST_TO_CONNECT` request with a matching UUID and pass-
phrase, it changes its state to `CONNECTED`.

The nRF24 PRX function itself does not provide a method to receive packets without acknowledging them. This would mean that if multiple *Configurator* are present, each listening for `TX_FREE_TO_CONNECT` packets, every *Configurator* will send ACK responses at the same time, causing issues.

To overcome this problem, the *Headless TX* broadcasts the `TX_FREE_TO_CONNECT` packets with the `W_TX_PAYLOAD_NOACK` feature of the nRF24. This method transmits a packet without requesting the PRX to acknowledge it. All *Configurator* can receive the `TX_FREE_TO_CONNECT` packet but no-one is acknowledging it.

In order to answer the `TX_FREE_TO_CONNECT` packet, immediately after the `TX_FREE_TO_CONNECT` was broadcast the *Headless TX* sends yet another packet with a single byte payload, but using the lower 5 Bytes of the UUID as address. This packet requests an acknowledgement from a PRX and is unique for each transmitter.

This means that when a *Configurator* sends a `CFG_REQUEST_TO_CONNECT` packet, it does so by configuring the address to the lower 5 Bytes of the UUID.


### nRF protocol *NOT CONNECTED* commands

Commands sent by the *Headless TX* always start with `TX_`. Command sent by the *Configurator* always start with `CFG_`.

All data is in little endian format, i.e. the least significant byte is transmitted first.

* `TX_FREE_TO_CONNECT`

    `0x30 uu uu uu uu uu uu uu uu uu tx tx tx tx tx tx tx tx tx tx tx tx tx tx tx tx b0 b1`

    uu:     Transmitter UUID 8 bytes
    tx:     Transmitter name 16 bytes (\0-terminated C-String [A-Z,a-z,0-9 ],
            padded with \0)
    b0, b1: Battery voltage in millivolts. Unsigned 16 bit integer.

    When sending this packet, the auto-acknowledge feature of the nRF24 is disabled. This is acheived by using the `W_TX_PAYLOAD_NOACK` nRF24 SPI command.

    Immediately after the packet was sent, another packet is sent, this time with the auto-acknowledge enabled (regular `W_TX_PAYLOAD` nRF SPI command). This command is sent on the address comprising of the lower 5 Bytes of the transmitter UUID. The payload is

    `0x30`

* `CFG_REQUEST_TO_CONNECT`

    `0x31 uu uu uu uu uu uu uu uu uu a0 a1 a2 a3 a4 p0 p1 oo sd`

    uu:     Transmitter UUID 8 bytes
    a0..4:  5 byte random nRF address to use during the connection
    p0..1:  unsigned 16-bit representation of the 4-digit pass-phrase
    oo, sd: offset and seed for a 7-bit Galois LFSR to calculate the hop channel sequence

    `oo` and `sd` are used to calculate the hop sequence for the configuration session.

    `sd` is the initial state of a [7-bit Galois LFSR](https://en.wikipedia.org/wiki/Linear-feedback_shift_register), `polynomial x^7 + x^6 + 1`. `sd` must be between 1 and 127.
    Offset `oo` is added after each LFSR round, with a modulo of 127.
    Values larger than 69 are skipped so that the resulting hop channels are between 1 and 69 only.

    The following pseudo-code shows how the 20 hop channels used during the *CONNECTED* state are calculated:

        oo = received_oo
        lfsr = received_sd

        function next
            do
                lsb = lfsr & 1;
                lfsr >>= 1;
                if (lsb) {
                    lfsr ^= 0x60;   // x^7 + x^6 + 1
                }
                channel = (lfsr + oo) % 127;
            while (channel > 69  &&  channel in hop_channels);
            // Note: this loop runs worst-case 7 times

            return channel
        end function next

        for i in 0..19
            hop_channel[i] = next()


### nRF protocol *CONNECTED* state

In *CONNECTED* state, the *Headless TX* sends one packet every 5 ms on the address received in the `CFG_REQUEST_TO_CONNECT` packet, looping through the 20 hop channels derived as described in the previous section.

Note that the first packet after connection is sent on the first hop channel. The *Configurator* knows when to expect the packet: ~5 ms after the `CFG_REQUEST_TO_CONNECT` was taken out of its TX buffer by the auto-acknowledge function.

Due to the use of the AA feature of the Shockburst protocol, every packet sent by the *Headless TX* is acknowledged by the *Configurator*, assuming the RF connection is stable and the hop sequence is synchronized.

If the *Headless TX* does not receive any acknowledgment within a window of 600 ms, it enters the *NOT CONNECTED* state.

If the *Configurator* does not receive any packet from from the *Headless TX* within a window of 600 ms, it enters the *NOT CONNECTED* state. There may be two reasons why the *Configurator* never received any packets:

- The `CFG_REQUEST_TO_CONNECT` message was lost or damaged during transmission.
- The pass-phrase was incorrect.

The *Configurator* may want to prompt the user for a new pass-phrase in case it tried a pass-phrase automatically -- for example a pass-phrase stored in its database.

While connected, the *Configurator* can issue requests to the *Headless TX* by adding the appropriate command to the auto-acknowledge payload.

When the *Headless TX* receives a packet acknowledgment without payload, or there was no acknowledgment for whatever reasons, the *Headless TX* sends an `TX_INFO` command as next packet.

The protocol is designed in such a way that every request from the *Configurator* (except `CFG_DISCONNECT`) causes the *Headless TX* to send a packet that is **not** `TX_INFO` as the next packet. This way the *Configurator* can know whether its request has been successfully received: if it is `TX_INFO`, something went wrong and the *Configurator* must resend the request.

The *Headless TX* does not resend packets. Info packets are repeated continuously, so no issue there. Missing responses to *Configurator* requests are handled by the *Configurator*.

When the *Headless TX* receives a payload along with the packet acknowledgment, it processes the command and responds with an appropriate answer in the next packet (sent during the next transmission burst; i.e. approx 5 ms later).


### nRF protocol *CONNECTED* commands


configurator/web-app/app/modules/config.js
var CONFIG_VERSIONS = {
    1: {
        // Legend:
        // o: offset with the parent structure
        // s: element size (number of bytes)
        // c: count (number of elements)
        // h: human-friendly name (optional)
        // t: type
        //   'u': unsigned integer
        //   'i': signed integer
        //   'c': string (Note: may not be 0 terminated if it fills the element)
        //   'uuid': 64-bit (8 bytes) universally unique identifier
        //   'CONFIG': schema describes the overall config structure
        //   'TX': schema describes a transmitter
        //   'MODEL': schema describes a model
        //   any other value: refers to named elements in TYPES[]

        CONFIG: { t: 'CONFIG', o: 0, s: 4476, c: 1,
            VERSION: {o: 0, s: 4, c: 1, t: 'u'},
            TX: {o: 4, s: 1588, c: 1, t: 's'},
            MODEL: {o: 1592, s: 2884, c: 1, t: 's'},
        },

        TX: { o: 4, s: 1588, c: 1, t: 'TX',
            UUID: {t: 'uuid', o: 0, s: 1, c: 8},
            NAME: {t: 'c', o: 8, s: 1, c: 16},
            LAST_CHANGED: {t: 'u', o: 24, s: 4, c: 1},
            TAG: {t: 'u', o: 28, s: 4, c: 1},
            HARDWARE_INPUTS: {t: 's', o: 32, s: 28, c: 32},
            HARDWARE_INPUTS_PCB_INPUT: {t: 's', o: 32, s: 20, c: 1},
            HARDWARE_INPUTS_PCB_INPUT_GPIOPORT: {t: 'u', o: 32, s: 4, c: 1},
            HARDWARE_INPUTS_PCB_INPUT_GPIO: {t: 'u', o: 36, s: 2, c: 1},
            HARDWARE_INPUTS_PCB_INPUT_ADC_CHANNEL: {t: 'u', o: 38, s: 1, c: 1},
            HARDWARE_INPUTS_PCB_INPUT_TYPE: {t: 'pcb_input_type_t', o: 39, s: 1, c: 1},
            HARDWARE_INPUTS_PCB_INPUT_PIN_NAME: {t: 'c', o: 40, s: 1, c: 10},
            HARDWARE_INPUTS_TYPE: {t: 'hardware_input_type_t', o: 52, s: 1, c: 1},
            HARDWARE_INPUTS_CALIBRATION: {t: 'u', o: 54, s: 2, c: 3},
            LOGICAL_INPUTS: {t: 's', o: 928, s: 20, c: 32},
            LOGICAL_INPUTS_TYPE: {t: 'input_type_t', h: 'Input type', o: 928, s: 1, c: 1},
            LOGICAL_INPUTS_SUB_TYPE: {t: 'input_sub_type_t', h: 'Push-button behavior', o: 929, s: 1, c: 1},
            LOGICAL_INPUTS_POSITION_COUNT: {t: 'u', o: 930, s: 1, c: 1},
            LOGICAL_INPUTS_HARDWARE_INPUTS: {t: 'u', h: 'Hardware inputs', o: 931, s: 1, c: 12},
            LOGICAL_INPUTS_LABELS: {t: 'input_label_t', h: 'Input labels', o: 943, s: 1, c: 5},
            TRIM_RANGE: {t: 'i', o: 1568, s: 4, c: 1},
            TRIM_STEP_SIZE: {t: 'i', o: 1572, s: 4, c: 1},
            BIND_TIMEOUT_MS: {t: 'u', o: 1576, s: 4, c: 1},
            DOUBLE_CLICK_TIMEOUT_MS: {t: 'u', o: 1580, s: 4, c: 1},
            PASSPHRASE: {t: 'u', o: 1584, s: 2, c: 1},
            LED_PWM_PERCENT: {t: 'u', o: 1586, s: 1, c: 1},
        },

        MODEL: { o: 1592, s: 2884, c: 1, t: 'MODEL',
            UUID: {t: 'uuid', o: 0, s: 1, c: 8},
            NAME: {t: 'c', o: 8, s: 1, c: 16},
            LAST_CHANGED: {t: 'u', o: 24, s: 4, c: 1},
            TAG: {t: 'u', o: 28, s: 4, c: 1},
            MIXER_UNITS: {t: 's', o: 32, s: 26, c: 100},
            MIXER_UNITS_CURVE: {t: 's', o: 32, s: 15, c: 1},
            MIXER_UNITS_CURVE_TYPE: {t: 'curve_type_t', o: 32, s: 1, c: 1},
            MIXER_UNITS_CURVE_SMOOTHING: {t: 'interpolation_type_t', o: 33, s: 1, c: 1},
            MIXER_UNITS_CURVE_POINTS: {t: 'i', o: 34, s: 1, c: 13},
            MIXER_UNITS_SRC: {t: 'src_label_t', h: 'Source', o: 47, s: 1, c: 1},
            MIXER_UNITS_DST: {t: 'channel_label_t', h: 'Destination', o: 48, s: 1, c: 1},
            MIXER_UNITS_SW: {t: 's', o: 49, s: 3, c: 1},
            MIXER_UNITS_SW_SW: {t: 'switch_label_t', h: 'Switch', o: 49, s: 1, c: 1},
            MIXER_UNITS_SW_CMP: {t: 'comparison_t', h: 'Comparison', o: 50, s: 1, c: 1},
            MIXER_UNITS_SW_VALUE: {t: 'u', o: 51, s: 1, c: 1},
            MIXER_UNITS_OP: {t: 'operation_type_t', h: 'Operation', o: 52, s: 1, c: 1},
            MIXER_UNITS_SCALAR: {t: 'i', o: 53, s: 1, c: 1},
            MIXER_UNITS_OFFSET: {t: 'i', o: 54, s: 1, c: 1},
            MIXER_UNITS_TAG: {t: 'u', o: 57, s: 1, c: 1},
            MIXER_UNITS_INVERT_SOURCE: {t: 'u', o: 55, s: 1, c: 1},
            MIXER_UNITS_APPLY_TRIM: {t: 'u', o: 56, s: 1, c: 1},
            LIMITS: {t: 's', o: 2632, s: 28, c: 8},
            LIMITS_EP_L: {t: 'i', o: 2632, s: 4, c: 1},
            LIMITS_EP_H: {t: 'i', o: 2636, s: 4, c: 1},
            LIMITS_SUBTRIM: {t: 'i', o: 2640, s: 4, c: 1},
            LIMITS_LIMIT_L: {t: 'i', o: 2644, s: 4, c: 1},
            LIMITS_LIMIT_H: {t: 'i', o: 2648, s: 4, c: 1},
            LIMITS_FAILSAFE: {t: 'i', o: 2652, s: 4, c: 1},
            LIMITS_SPEED: {t: 'u', o: 2656, s: 1, c: 1},
            LIMITS_INVERT: {t: 'u', o: 2657, s: 1, c: 1},
            RF_PROTOCOL_TYPE: {t: 'rf_protocol_type_t', o: 2856, s: 1, c: 1},
            RF: {t: 's', o: 2857, s: 25, c: 1},
            RF_PROTOCOL_HK310: {t: 's', o: 2857, s: 25, c: 1},
            RF_PROTOCOL_HK310_HOP_CHANNELS: {t: 'u', o: 2857, s: 1, c: 20},
            RF_PROTOCOL_HK310_ADDRESS: {t: 'u', o: 2877, s: 1, c: 5},
        },

        TYPES: {
            pcb_input_type_t: {
                'Analog/Digital': 1,
                'Digital': 2,
            },
            hardware_input_type_t: {
                'Input not used': 0,
                'Analog, returns to center': 1,
                'Analog, center detent': 2,
                'Analog': 3,
                'Analog, positive only': 4,
                'On/Off switch': 5,
                'On/Off/On switch': 6,
                'Push-button': 7,
            },
            hardware_input_type_t_digital: {
                'Input not used': 0,
                'On/Off switch': 5,
                'On/Off/On switch': 6,
                'Push-button': 7,
            },
            input_type_t: {
                'Analog': 1,
                'Switch': 2,
                'BCD switch': 3,
                'Momentary switch': 4,
                'Trim': 5,
            },
            input_sub_type_t: {
                'Up/Down buttons': 1,
                'Increment-and-loop': 2,
                'Decrement-and-loop': 3,
                'Sawtooth': 4,
                'Double-click for decrement': 5,
            },
            input_label_t: {
                'ST': 1,
                'TH': 2,
                'THR': 3,
                'RUD': 4,
                'AIL': 5,
                'ELE': 6,
                'AUX': 7,
                'ST-DR': 8,
                'RUD-DR': 9,
                'AIL-DR': 10,
                'ELE-DR': 11,
                'TH-DR': 12,
                'THR-DR': 13,
                'TH-HOLD': 14,
                'GEAR': 15,
                'FLAPS': 16,
                'TRAINER': 17,
                'SIDE-L': 18,
                'SIDE-R': 19,
                'POT1': 20,
                'POT2': 21,
                'POT3': 22,
                'POT4': 23,
                'POT5': 24,
                'POT6': 25,
                'POT7': 26,
                'POT8': 27,
                'POT9': 28,
                'SW1': 29,
                'SW2': 30,
                'SW3': 31,
                'SW4': 32,
                'SW5': 33,
                'SW7': 35,
                'SW8': 36,
                'SW9': 37,
            },
            switch_label_t: {
                'NONE': 0,
                'ST': 1,
                'TH': 2,
                'THR': 3,
                'RUD': 4,
                'AIL': 5,
                'ELE': 6,
                'AUX': 7,
                'ST-DR': 8,
                'RUD-DR': 9,
                'AIL-DR': 10,
                'ELE-DR': 11,
                'TH-DR': 12,
                'THR-DR': 13,
                'TH-HOLD': 14,
                'GEAR': 15,
                'FLAPS': 16,
                'TRAINER': 17,
                'SIDE-L': 18,
                'SIDE-R': 19,
                'POT1': 20,
                'POT2': 21,
                'POT3': 22,
                'POT4': 23,
                'POT5': 24,
                'POT6': 25,
                'POT7': 26,
                'POT8': 27,
                'POT9': 28,
                'SW1': 29,
                'SW2': 30,
                'SW3': 31,
                'SW4': 32,
                'SW5': 33,
                'SW7': 35,
                'SW8': 36,
                'SW9': 37,
            },
            channel_label_t: {
                'CH1': 0,
                'CH2': 1,
                'CH3': 2,
                'CH4': 3,
                'CH5': 4,
                'CH6': 5,
                'CH7': 6,
                'CH8': 7,
                'VIRTUAL1': 8,
                'VIRTUAL2': 9,
                'VIRTUAL3': 10,
                'VIRTUAL4': 11,
                'VIRTUAL5': 12,
                'VIRTUAL6': 13,
                'VIRTUAL7': 14,
                'VIRTUAL8': 15,
                'VIRTUAL9': 16,
                'VIRTUAL10': 17,
            },
            src_label_t: {
                'ST': 2,
                'TH': 3,
                'THR': 4,
                'RUD': 5,
                'AIL': 6,
                'ELE': 7,
                'AUX': 8,
                'ST-DR': 9,
                'RUD-DR': 10,
                'AIL-DR': 11,
                'ELE-DR': 12,
                'TH-DR': 13,
                'THR-DR': 14,
                'TH-HOLD': 15,
                'GEAR': 16,
                'FLAPS': 17,
                'TRAINER': 18,
                'SIDE-L': 19,
                'SIDE-R': 20,
                'POT1': 21,
                'POT2': 22,
                'POT3': 23,
                'POT4': 24,
                'POT5': 25,
                'POT6': 26,
                'POT7': 27,
                'POT8': 28,
                'POT9': 29,
                'SW1': 30,
                'SW2': 31,
                'SW3': 32,
                'SW4': 33,
                'SW5': 34,
                'SW7': 36,
                'SW8': 37,
                'SW9': 38,
                'CH1': 39,
                'CH2': 40,
                'CH3': 41,
                'CH4': 42,
                'CH5': 43,
                'CH6': 44,
                'CH7': 45,
                'CH8': 46,
                'VIRTUAL1': 47,
                'VIRTUAL2': 48,
                'VIRTUAL3': 49,
                'VIRTUAL4': 50,
                'VIRTUAL5': 51,
                'VIRTUAL6': 52,
                'VIRTUAL7': 53,
                'VIRTUAL8': 54,
                'VIRTUAL9': 55,
                'VIRTUAL10': 56,
            },
            live_t: {
                'ST': 1,
                'TH': 2,
                'THR': 3,
                'RUD': 4,
                'AIL': 5,
                'ELE': 6,
                'AUX': 7,
                'ST-DR': 8,
                'RUD-DR': 9,
                'AIL-DR': 10,
                'ELE-DR': 11,
                'TH-DR': 12,
                'THR-DR': 13,
                'TH-HOLD': 14,
                'GEAR': 15,
                'FLAPS': 16,
                'TRAINER': 17,
                'SIDE-L': 18,
                'SIDE-R': 19,
                'POT1': 20,
                'POT2': 21,
                'POT3': 22,
                'POT4': 23,
                'POT5': 24,
                'POT6': 25,
                'POT7': 26,
                'POT8': 27,
                'POT9': 28,
                'SW1': 29,
                'SW2': 30,
                'SW3': 31,
                'SW4': 32,
                'SW5': 33,
                'SW7': 35,
                'SW8': 36,
                'SW9': 37,
                'CH1': 38,
                'CH2': 39,
                'CH3': 40,
                'CH4': 41,
                'CH5': 42,
                'CH6': 43,
                'CH7': 44,
                'CH8': 45,
                'VIRTUAL1': 46,
                'VIRTUAL2': 47,
                'VIRTUAL3': 48,
                'VIRTUAL4': 49,
                'VIRTUAL5': 50,
                'VIRTUAL6': 51,
                'VIRTUAL7': 52,
                'VIRTUAL8': 53,
                'VIRTUAL9': 54,
                'VIRTUAL10': 55,
                'HIDDEN1': 56,
                'HIDDEN2': 57,
                'HIDDEN3': 58,
                'HIDDEN4': 59,
                'HIDDEN5': 60,
                'HIDDEN6': 61,
                'HIDDEN7': 62,
                'HIDDEN8': 63,
                'HIDDEN9': 64,
                'HIDDEN10': 65,
                'HIDDEN11': 66,
                'HIDDEN12': 67,
                'HIDDEN13': 68,
                'HIDDEN14': 69,
                'HIDDEN15': 70,
                'HIDDEN16': 71,
                'HIDDEN17': 72,
                'HIDDEN18': 73,
                'HIDDEN19': 74,
                'HIDDEN20': 75,
                'HIDDEN21': 76,
                'HIDDEN22': 77,
                'HIDDEN23': 78,
                'HIDDEN24': 79,
                'HIDDEN25': 80,
                'HIDDEN26': 81,
                'HIDDEN27': 82,
                'HIDDEN28': 83,
                'HIDDEN29': 84,
                'HIDDEN30': 85,
                'HIDDEN31': 86,
                'HIDDEN32': 87,
                'HIDDEN33': 88,
                'HIDDEN34': 89,
                'HIDDEN35': 90,
                'HIDDEN36': 91,
                'HIDDEN37': 92,
                'HIDDEN38': 93,
                'HIDDEN39': 94,
                'HIDDEN40': 95,
                'HIDDEN41': 96,
                'HIDDEN42': 97,
                'HIDDEN43': 98,
                'HIDDEN44': 99,
                'HIDDEN45': 100,
                'HIDDEN46': 101,
                'HIDDEN47': 102,
                'HIDDEN48': 103,
                'HIDDEN49': 104,
                'HIDDEN50': 105,
                'RF-CH1': 106,
                'RF-CH2': 107,
                'RF-CH3': 108,
                'RF-CH4': 109,
                'RF-CH5': 110,
                'RF-CH6': 111,
                'RF-CH7': 112,
                'RF-CH8': 113,
                'Battery (mV)': 114,
                'ADC0 (raw)': 115,
                'ADC1 (raw)': 116,
                'ADC2 (raw)': 117,
                'ADC3 (raw)': 118,
                'ADC4 (raw)': 119,
                'ADC5 (raw)': 120,
                'ADC6 (raw)': 121,
                'ADC7 (raw)': 122,
                'ADC8 (raw)': 123,
                'ADC9 (raw)': 124,
                'ADC10 (raw)': 125,
                'ADC11 (raw)': 126,
                'ADC12 (raw)': 127,
                'ADC13 (raw)': 128,
                'ADC14 (raw)': 129,
                'ADC15 (raw)': 130,
                'ADC16 (raw)': 131,
                'ADC17 (raw)': 132,
                'ADC18 (raw)': 133,
                'ADC19 (raw)': 134,
                'ADC20 (raw)': 135,
                'ADC21 (raw)': 136,
                'ADC22 (raw)': 137,
                'ADC23 (raw)': 138,
                'ADC24 (raw)': 139,
                'ADC25 (raw)': 140,
                'ADC26 (raw)': 141,
                'ADC27 (raw)': 142,
                'ADC28 (raw)': 143,
                'ADC29 (raw)': 144,
                'ADC30 (raw)': 145,
                'ADC31 (raw)': 146,
            },
            rf_protocol_type_t: {
                'HobbyKing HKR3000': 0,
            },
            operation_type_t: {
                '=': 0,
                '+': 1,
                '*': 2,
                'MIN': 3,
                'MAX': 4,
            },
            comparison_t: {
                '==': 0,
                '!=': 1,
                '>': 2,
                '>=': 3,
                '<': 4,
                '<=': 5,
            },
            curve_type_t: {
                'Linear': 0,
                'Fixed value': 1,
                'Min/Max': 2,
                '0/Max': 3,
                '>0': 4,
                '<0': 5,
                'Absolute': 6,
                'Expo': 7,
                'Deadband': 8,
                '3-Point': 9,
                '5-Point': 10,
                '7-Point': 11,
                '9-Point': 12,
                '11-Point': 13,
                '13-Point': 14,
            },
            interpolation_type_t: {
                'Linear': 0,
                'Smoothing': 1,
            },
            curve_points_t: {
                "Linear": [],
                "Fixed value": ["Value"],
                "Min/Max": ["Threshold"],
                "0/Max": ["Threshold"],
                ">0": ["Threshold"],
                "<0": ["Threshold"],
                "Absolute": ["Threshold"],
                "Expo": [">0", "<0"],
                "Deadband": [">0", "<0"],
                "3-Point": ["+100%", "0", "-100%"],
                "5-Point": ["+100%", "+50%", "0", "-50%", "-100%"],
                "7-Point": ["+100%", "+66%", "+33%", "0", "-33%", "-66%", "-100%"],
                "9-Point": ["+100%", "+75%", "+50%", "+25%", "0", "-25", "-50%", "-75%", "-100%"],
                "11-Point": ["+100%", "+80%", "+60%", "+40%", "+20%", "0", "-20%", "-40%", "-60%", "-80%", "-100%"],
                "13-Point": ["+100%", "+83%", "+67%", "+50%", "+33%", "+17%", "0", "-17%", "-33%", "-50%", "-67%", "-83%", "-100%"],
            },
        }
    }
};

window['CONFIG_VERSIONS'] = CONFIG_VERSIONS;
